tests/python_package_test/test_engine.py

```diff
@@ -828,3 +829,30 @@ def preprocess_data(dtrain, dtest, params):
         results = lgb.cv(params, dataset, num_boost_round=10, fpreproc=preprocess_data)
         self.assertIn('multi_logloss-mean', results)
         self.assertEqual(len(results['multi_logloss-mean']), 10)
+
+    @unittest.skipIf(psutil.virtual_memory().available / 1024 / 1024 / 1024 < 3, 'not enough RAM')
+    def test_model_size(self):
+        X, y = load_boston(True)
+        data = lgb.Dataset(X, y)
+        bst = lgb.train({'verbose': -1}, data, num_boost_round=2)
+        y_pred = bst.predict(X)
+        model_str = bst.model_to_string()
+        one_tree = model_str[model_str.find('Tree=1'):model_str.find('end of trees')]
+        one_tree_size = len(one_tree)
+        one_tree = one_tree.replace('Tree=1', 'Tree={}')
+        multiplier = int(2**31 / (one_tree_size + len(str(one_tree_size)) + 1))
```

Comment:
@guolinke Now test fails not with `MemoryError`, but with `ci/test.sh: line 120:  4299 Killed  pytest $BUILD_DIRECTORY/tests`.
Should I add a magic number here to take into account the growing size of tree(due to Tree=1,...100,...,10000) and decrease the difference between actual model size and desired? Or is any other easy way to calculate the needed number of trees?

Reply:
actually, the tree_size field could be deleted directly. This will slow down the model loading, but I think it is okay for test. 
BTW, I think the simplest solution is padding many `\r\n`, as it will increase the length of model string and will be ignored by model parsing.

Reply:
Oh, `tree_sizes` can be simply removed, it's great!

Now, some fixed little number of trees (100) is appended to the model and then `\n`-padding is used to reach 2**31 total model size. Unfortunately, it causes weird memory consumption behaviour during the model loading process, please see the screenshot below:

![image](https://user-images.githubusercontent.com/25141164/51775277-4143e600-2106-11e9-8ef5-6d87513dc87e.png)

Compare with simply huge number of trees without `\n`-padding behaviour (`multiplier=1370000` instead of 100, and the total model size is even bigger compared with `\n`-padding):

![image](https://user-images.githubusercontent.com/25141164/51775283-486af400-2106-11e9-90ad-4dbdb6e052db.png)

Reply:
That is weird, it should be in this loop...
https://github.com/Microsoft/LightGBM/blob/b37065dbd5510079aac341a34895f4854c782fcb/src/boosting/gbdt_model_text.cpp#L442-L459

Reply:
I tried to append '\n's at the model's end - got the same weird behavior.

Reply:
will this cause memory error in local ?

Reply:
Nope! On my home machine it consumed all 32GB in last peak and then released all excess memory; after that model was loaded successfully.

Actually, `\n` padding at the end is inappropriate action because we have the following code, which checks only the last and one before last lines for pandas section:
https://github.com/Microsoft/LightGBM/blob/b96849918d802f750657b3dc3d17324e070993a2/python-package/lightgbm/basic.py#L318-L323

Also, I tried to move `cur_line` under the `if` statement, but it didn't help.
https://github.com/Microsoft/LightGBM/blob/b96849918d802f750657b3dc3d17324e070993a2/src/boosting/gbdt_model_text.cpp#L445-L446

Reply:
I tried to use spaces instead of new lines. The situation became much better.

```diff
index 049c682..b06ce16 100644
--- a/tests/python_package_test/test_engine.py
+++ b/tests/python_package_test/test_engine.py
@@ -843,7 +843,7 @@ class TestEngine(unittest.TestCase):
         multiplier = 100
         total_trees = multiplier + 2
         new_model_str = (model_str[:model_str.find('tree_sizes')]
-                         + '\n' * (2**31 - one_tree_size * total_trees)
+                         + '\n\n\n' + ' ' * (2**31 - one_tree_size * total_trees)
                          + model_str[model_str.find('Tree=0'):model_str.find('end of trees')]
                          + (one_tree * multiplier).format(*range(2, total_trees))
                          + model_str[model_str.find('end of trees'):])
```

![image](https://user-images.githubusercontent.com/25141164/51794688-72e1ad80-21e9-11e9-9340-e2f818df4a2c.png)

But surprisingly it loaded 101 trees instead of 102...

After that I tried to move whitespaces to model's end: everything became even much better and the correct number of trees was loaded.

```diff
index 049c682..31743b8 100644
--- a/tests/python_package_test/test_engine.py
+++ b/tests/python_package_test/test_engine.py
@@ -843,10 +843,11 @@ class TestEngine(unittest.TestCase):
         multiplier = 100
         total_trees = multiplier + 2
         new_model_str = (model_str[:model_str.find('tree_sizes')]
-                         + '\n' * (2**31 - one_tree_size * total_trees)
+                         + '\n\n'
                          + model_str[model_str.find('Tree=0'):model_str.find('end of trees')]
                          + (one_tree * multiplier).format(*range(2, total_trees))
-                         + model_str[model_str.find('end of trees'):])
+                         + model_str[model_str.find('end of trees'):]
+                         + ' ' * (2**31 - one_tree_size * total_trees))
         self.assertGreater(len(new_model_str), 2**31)
         bst.model_from_string(new_model_str, verbose=False)
         self.assertEqual(bst.num_trees(), total_trees)
```

![image](https://user-images.githubusercontent.com/25141164/51794710-d79d0800-21e9-11e9-8c4a-dc0951141316.png)

Reply:
great! can this pass the test now?

Reply:
Haven't tried yet.

BTW, seems that this line is the root cause of weird memory consumption.
https://github.com/Microsoft/LightGBM/blob/b96849918d802f750657b3dc3d17324e070993a2/python-package/lightgbm/basic.py#L318


![image](https://user-images.githubusercontent.com/25141164/51801947-862a6280-2255-11e9-916c-ac79da4a7427.png)
![image](https://user-images.githubusercontent.com/25141164/51801950-8e829d80-2255-11e9-89da-60eaaa4411b2.png)

Don't you know why spaces in the middle of the model cause wrong number of trees are being loaded?
> But surprisingly it loaded 101 trees instead of 102...

Reply:
because it will break here:
https://github.com/Microsoft/LightGBM/blob/b37065dbd5510079aac341a34895f4854c782fcb/src/boosting/gbdt_model_text.cpp#L455
maybe we should remove this break in cpp side.

Reply:
Oh, got it! Thanks!

> maybe we should remove this break in cpp side.

Not sure that it's actually needed. Unexpected whitespaces in a model file are not something that can happen during save/load use cases, and manual editing is not something that we're supporting. That question was just my curiosity.